test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs

```diff
@@ -0,0 +1,211 @@
+
+        public bool IsTopLevel { get; set; } = true;
+
+        public string Type { get; set; }
+
+        public string ResourceName { get; set; }
+
+        public string OperationName { get; set; }
+
+        public string ServiceName { get; set; }
+
+        public virtual string Detail()
+        {
+            return $"service={ServiceName}, operation={OperationName}, type={Type}, resource={ResourceName}";
+        }
+
+        /// <summary>
+        /// Override for custom filters.
+        /// </summary>
+        /// <param name="span">The span on which to filter.</param>
+        /// <returns>Whether the span qualifies for this expectation.</returns>
+        public virtual bool ShouldInspect(MockTracerAgent.Span span)
+        {
+            return span.Service == ServiceName
+                && span.Name == OperationName
+                && span.Type == Type;
+        }
+
+        /// <summary>
+        /// The aggregate assertion which is run for a test.
+        /// </summary>
+        /// <param name="span">The span being asserted against.</param>
+        /// <param name="message">The developer friendly message for the test failure.</param>
+        /// <returns>Whether the span meets expectations.</returns>
+        public bool MeetsExpectations(MockTracerAgent.Span span, out string message)
+        {
+            message = string.Empty;
+
+            var messages = new List<string>();
+
+            foreach (var assertion in Assertions)
+            {
+                var mismatchMessage = assertion(span);
+                if (!string.IsNullOrWhiteSpace(mismatchMessage))
+                {
+                    messages.Add(mismatchMessage);
+                }
+            }
+
+            if (messages.Any())
+            {
+                message = string.Join(",", messages);
+                return false;
+            }
+
+            return true;
+        }
+
+        public void TagShouldExist(string tagKey, Func<MockTracerAgent.Span, bool> when)
+        {
+            Assertions.Add(span =>
+            {
+                if (when(span) && !span.Tags.ContainsKey(tagKey))
+                {
+                    return $"Tag {tagKey} is missing from span.";
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterDelegateExpectation(Func<MockTracerAgent.Span, IEnumerable<string>> expectation)
+        {
+            Assertions.Add(span =>
+            {
+                var failures = expectation(span);
+
+                if (failures.Any())
+                {
+                    return string.Join(",", failures);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterCustomExpectation(string keyForMessage, Func<MockTracerAgent.Span, string> actual, Func<MockTracerAgent.Span, string> expected)
```

Comment:
Could we simplify this so that the `expected` param is not a Func but instead just a string that we compare against?

Reply:
Sure, we can always change it back if we need to run actual logic, or I suppose the delegate one could be used.